File: sketchybar/top/shortcuts_logic.lua
-- Pure skhdrc → shortcut-list parser (no sbar). Turns binding lines + their
-- preceding comments into sections of { desc, keys } rows for the popup.
-- Tested by sketchybar/top/tests/shortcuts_logic_test.lua.

local logic = {}

local MOD_ORDER = { "fn", "ctrl", "alt", "shift", "cmd" }
local MOD_GLYPH = { fn = "fn", ctrl = "⌃", alt = "⌥", shift = "⇧", cmd = "⌘" }
local MOD_EXPAND = {
  meh = { "ctrl", "alt", "shift" },
  hyper = { "ctrl", "alt", "shift", "cmd" },
}
local MOD_WORDS = { fn = true, ctrl = true, alt = true, shift = true, cmd = true }
for word in pairs(MOD_EXPAND) do
  MOD_WORDS[word] = true
end

local KEY_GLYPH = {
  space = "Space",
  left = "←",
  right = "→",
  up = "↑",
  down = "↓",
  ["return"] = "↩",
  tab = "⇥",
  escape = "⎋",
  delete = "⌫",
}

local function trim(s)
  return (s:gsub("^%s+", ""):gsub("%s+$", ""))
end

local function capitalize(s)
  return (s:gsub("^%l", string.upper))
end

-- Comment content that is pure box/banner decoration (##### or =====).
local function is_decoration(content)
  return content:match "^[#=%-%s]*$" ~= nil
end

-- "# fn + alt - n : yabai ..." — a commented-out binding, not a description.
local function is_disabled_binding(content)
  local first = content:match "^(%a+)"
  if not first or not MOD_WORDS[first:lower()] then
    return false
  end
  return content:match "%-%s*%S+%s*:" ~= nil
end

-- Parse "fn + ctrl - y" → glyph chord prefix ("fn⌃") or nil when invalid.
local function parse_mods(modstr)
  local present = {}
  for token in modstr:gmatch "[^%+%s]+" do
    local word = token:lower()
    if MOD_EXPAND[word] then
      for _, m in ipairs(MOD_EXPAND[word]) do
        present[m] = true
      end
    elseif MOD_GLYPH[word] then
      present[word] = true
    else
      return nil
    end
  end
  local out = {}
  for _, m in ipairs(MOD_ORDER) do
    if present[m] then
      out[#out + 1] = MOD_GLYPH[m]
    end
  end
  return table.concat(out)
end

local function parse_binding(line)
  local lhs, cmd = line:match "^([^:]+):%s*(.*)$"
  if not lhs then
    return nil
  end
  local modstr, key = lhs:match "^(.-)%-%s*(%S+)%s*$"
  if not modstr then
    return nil
  end
  local mods = parse_mods(modstr)
  if not mods or cmd == "" then
    return nil
  end
  return { mods = mods, key = key:lower(), cmd = trim(cmd) }
end

-- Strip implementation noise from a comment: "(u)p" hints, "— …" and
-- "(signal …)" / ": …" tails. Keeps the human-facing half.
function logic.pretty_desc(s)
  s = trim(s)
  s = s:gsub("%((%a)%)", "%1")
  local dash = s:find("—", 1, true)
  if dash then
    s = s:sub(1, dash - 1)
  end
  local paren = s:find(" (", 1, true)
  if paren then
    s = s:sub(1, paren - 1)
  end
  local colon = s:find(": ", 1, true)
  if colon then
    s = s:sub(1, colon - 1)
  end
  s = trim(s):gsub("[%.;,]+$", "")
  return capitalize(s)
end

-- Best-effort description for bindings that carry no comment.
function logic.derive_desc(cmd)
  local url = cmd:match "vicinae://launch/(%S+)"
  if url then
    url = url:gsub('["\']$', "")
    local words, seen = {}, {}
    for seg in url:gmatch "[^/]+" do
      if not seg:match "[@%.]" then
        for w in seg:gmatch "[^%-]+" do
          local lw = w:lower()
          if not seen[lw] then
            seen[lw] = true
            words[#words + 1] = lw
          end
        end
      end
    end
    if #words == 0 then
      words = { (url:match "([^/]+)/?$" or url):lower() }
    end
    return capitalize(table.concat(words, " "))
  end
  if cmd:match "systempreferences" then
    return "System Settings"
  end
  local app = cmd:match 'open%s+%-a%s+"([^"]+)"' or cmd:match "open%s+%-a%s+([%w%.]+)"
  if app then
    return capitalize(app)
  end
  app = cmd:match '/Applications/([^/"]+)%.app'
  if app then
    return app
  end
  local arg = cmd:match 'toggle_app%.bash"%s*"([^"]+)"'
  if arg then
    local var = arg:match "^%$([%u_]+)$"
    if var then
      arg = capitalize(var:gsub("_NAME$", ""):lower())
    end
    return "Toggle " .. arg
  end
  local script = cmd:match 'bash%s+"[^"]*/([%w%-_]+)"'
  if script then
    return capitalize(script:gsub("%-", " "))
  end
  if cmd:match "^yabai %-m space %-%-focus %d" then
    return "Focus space"
  end
  return capitalize(cmd:match "^(%S+)" or cmd)
end

local function display_key(key)
  if KEY_GLYPH[key] then
    return KEY_GLYPH[key]
  end
  return key:upper()
end

-- "1".."9" consecutive → "1–9"; else "H/J/K/L".
local function display_keys(keys)
  if #keys > 1 then
    local nums = {}
    for i, k in ipairs(keys) do
      nums[i] = k:match "^%d$" and tonumber(k) or nil
    end
    local consecutive = nums[1] ~= nil
    for i = 2, #keys do
      if not nums[i] or nums[i] ~= nums[i - 1] + 1 then
        consecutive = false
        break
      end
    end
    if consecutive then
      return keys[1] .. "–" .. keys[#keys]
    end
  end
  local out = {}
  for i, k in ipairs(keys) do
    out[i] = display_key(k)
  end
  return table.concat(out, "/")
end

local function chord(mods, keys)
  local k = display_keys(keys)
  if mods == "" then
    return k
  end
  return mods .. " " .. k
end

-- Group consecutive entries that describe one action, resolve descriptions,
-- then merge adjacent groups that ended up identical (fn 1–5 + fn 6–9).
local function build_rows(entries)
  local groups = {}
  for _, e in ipairs(entries) do
    local norm = e.cmd:gsub("%d+", "#")
    local prev = groups[#groups]
    local joins = prev
      and prev.mods == e.mods
      and (
        (e.desc ~= nil and e.desc == prev.desc)
        or (e.desc == nil and prev.desc == nil and norm == prev.norm)
      )
    if joins then
      prev.keys[#prev.keys + 1] = e.key
    else
      groups[#groups + 1] = { mods = e.mods, keys = { e.key }, desc = e.desc, norm = norm, cmd = e.cmd }
    end
  end
  for _, g in ipairs(groups) do
    g.text = g.desc and logic.pretty_desc(g.desc) or logic.derive_desc(g.cmd)
  end
  local merged = {}
  for _, g in ipairs(groups) do
    local prev = merged[#merged]
    if prev and prev.mods == g.mods and prev.text == g.text then
      for _, k in ipairs(g.keys) do
        prev.keys[#prev.keys + 1] = k
      end
    else
      merged[#merged + 1] = g
    end
  end
  local rows = {}
  for _, g in ipairs(merged) do
    rows[#rows + 1] = { desc = g.text, keys = chord(g.mods, g.keys) }
  end
  return rows
end

-- text → { { title, rows = { { desc, keys } } } }
function logic.parse(text)
  -- Join continuation lines (trailing backslash) into logical lines.
  local lines = {}
  local acc = nil
  for line in (text .. "\n"):gmatch "(.-)\n" do
    acc = acc and (acc .. " " .. trim(line)) or line
    if acc:match "\\%s*$" then
      acc = acc:gsub("\\%s*$", "")
    else
      lines[#lines + 1] = acc
      acc = nil
    end
  end

  local sections = {}
  local current = { title = "General", entries = {} }
  local pending, last_desc, last_mods = nil, nil, nil

  local function flush()
    if #current.entries > 0 then
      sections[#sections + 1] = current
    end
  end

  local function comment_content(line)
    local t = trim(line)
    if t:sub(1, 1) ~= "#" then
      return nil
    end
    -- Strip ALL leading hash runs ("# # comment" → "comment").
    return t:match "^[#%s]*(.-)%s*$" or ""
  end

  local i = 1
  while i <= #lines do
    local line = trim(lines[i])
    if line == "" then
      pending, last_desc, last_mods = nil, nil, nil
    elseif line:sub(1, 2) == "#!" then
      -- shebang
    elseif line:sub(1, 1) == "#" then
      local content = comment_content(line)
      if is_decoration(content) then
        pending, last_desc, last_mods = nil, nil, nil
        -- Boxed section header: deco / "#  Title  #" / deco.
        local mid = lines[i + 1] and comment_content(lines[i + 1])
        local bot = lines[i + 2] and comment_content(lines[i + 2])
        if mid and bot and is_decoration(bot) then
          local title = trim(mid:gsub("%s*#+%s*$", ""))
          if title ~= "" and not is_decoration(title) and not title:find(":", 1, true) then
            flush()
            current = { title = title, entries = {} }
            i = i + 2
          end
        end
      elseif is_disabled_binding(content) then
        -- Commented-out binding: skip, keep surrounding comment context.
      else
        pending, last_desc, last_mods = content, nil, nil
      end
    else
      local entry = parse_binding(line)
      if entry then
        if pending then
          entry.desc = pending
          pending = nil
        elseif last_mods == entry.mods then
          entry.desc = last_desc
        end
        last_desc, last_mods = entry.desc, entry.mods
        current.entries[#current.entries + 1] = entry
      end
    end
    i = i + 1
  end
  flush()

  local out = {}
  for _, sec in ipairs(sections) do
    out[#out + 1] = { title = sec.title, rows = build_rows(sec.entries) }
  end
  return out
end

return logic

File: asahi/quickshell/remix/modules/bar/shortcuts_logic.js
// Hyprland bindings.lua → shortcut-list parser (no QML). Turns hl.bind /
// media_bind calls + their -- section comments into { title, rows } for the
// bar popup. Mirrors sketchybar/top/shortcuts_logic.lua.
// QML: import "shortcuts_logic.js" as ShortcutsLogic
// Node: require("./shortcuts_logic.js")

var MOD_ORDER = ["super", "ctrl", "alt", "shift"]
var MOD_GLYPH = { super: "⌘", ctrl: "⌃", alt: "⌥", shift: "⇧" }
var MOD_ALIASES = {
  super: "super",
  mainmod: "super",
  mod: "super",
  control: "ctrl",
  ctrl: "ctrl",
  alt: "alt",
  shift: "shift"
}

var KEY_GLYPH = {
  space: "Space",
  left: "←",
  right: "→",
  up: "↑",
  down: "↓",
  return: "↩",
  tab: "⇥",
  escape: "⎋",
  delete: "⌫",
  page_up: "⇞",
  page_down: "⇟",
  caps_lock: "Caps",
  mouse_down: "Scroll↓",
  mouse_up: "Scroll↑",
  "mouse:272": "LMB",
  "mouse:273": "RMB",
  xf86audioraisevolume: "Vol+",
  xf86audiolowervolume: "Vol-",
  xf86audiomute: "Mute",
  xf86audiomicmute: "Mic",
  xf86audioplay: "Play",
  xf86audiopause: "Pause",
  xf86audionext: "Next",
  xf86audioprev: "Prev",
  xf86monbrightnessup: "Bright+",
  xf86monbrightnessdown: "Bright-",
  xf86search: "Kbd+",
  xf86launcha: "Kbd-"
}

function trim(s) {
  return String(s || "").replace(/^\s+|\s+$/g, "")
}

function capitalize(s) {
  s = String(s || "")
  return s.replace(/^\s*\S/, function (ch) { return ch.toUpperCase() })
}

function prettyDesc(s) {
  s = trim(s)
  s = s.replace(/\(([A-Za-z])\)/g, "$1")
  var dash = s.indexOf("—")
  if (dash !== -1) s = s.slice(0, dash)
  var paren = s.indexOf(" (")
  if (paren !== -1) s = s.slice(0, paren)
  var colon = s.indexOf(": ")
  if (colon !== -1) s = s.slice(0, colon)
  s = trim(s).replace(/[.;,]+$/, "")
  return capitalize(s)
}

function deriveDesc(action) {
  var a = String(action || "")
  if (/workspace\s*=\s*"e\+1"/.test(a)) return "Next workspace"
  if (/workspace\s*=\s*"e-1"/.test(a)) return "Previous workspace"
  if (/window\.close/.test(a)) return "Close window"
  if (/window\.float/.test(a)) return "Toggle floating"
  if (/window\.center/.test(a)) return "Center floating window"
  if (/window\.pin/.test(a)) return "Toggle pin window"
  if (/window\.fullscreen/.test(a)) return "Toggle fullscreen"
  if (/window\.pseudo/.test(a)) return "Toggle pseudo"
  if (/group\.toggle/.test(a)) return "Toggle group"
  if (/layout\s*"togglesplit"/.test(a)) return "Toggle split"
  var script = a.match(/asahi-([a-z0-9-]+)/)
  if (script) return capitalize(script[1].replace(/-/g, " "))
  var ident = a.match(/dsp\.([a-zA-Z_]+)/)
  if (ident) return capitalize(ident[1].replace(/_/g, " "))
  return "Shortcut"
}

function displayKey(key) {
  var raw = String(key || "")
  var lower = raw.toLowerCase()
  if (KEY_GLYPH[lower]) return KEY_GLYPH[lower]
  if (/^[a-z]$/.test(lower)) return lower.toUpperCase()
  if (/^\d$/.test(raw)) return raw
  return raw
}

function displayKeys(keys) {
  if (keys.length > 1) {
    var nums = []
    var allNum = true
    for (var i = 0; i < keys.length; i++) {
      if (!/^\d$/.test(keys[i])) {
        allNum = false
        break
      }
      nums.push(Number(keys[i]))
    }
    if (allNum) {
      var consecutive = true
      for (var n = 1; n < nums.length; n++) {
        if (nums[n] !== nums[n - 1] + 1) {
          consecutive = false
          break
        }
      }
      if (consecutive) return keys[0] + "–" + keys[keys.length - 1]
      // Workspace 1–10 on Hyprland uses key 0 for workspace 10.
      if (nums.length >= 2 && nums[nums.length - 1] === 0 && nums[0] === 1) {
        var wrap = true
        for (var w = 1; w < nums.length - 1; w++) {
          if (nums[w] !== w + 1) {
            wrap = false
            break
          }
        }
        if (wrap) return keys[0] + "–" + keys[keys.length - 1]
      }
    }
  }
  var out = []
  for (var k = 0; k < keys.length; k++) out.push(displayKey(keys[k]))
  return out.join("/")
}

function formatMods(present) {
  var out = []
  for (var i = 0; i < MOD_ORDER.length; i++) {
    if (present[MOD_ORDER[i]]) out.push(MOD_GLYPH[MOD_ORDER[i]])
  }
  return out.join("")
}

function chord(mods, keys) {
  var k = displayKeys(keys)
  if (!mods) return k
  return mods + " " + k
}

function parseChord(expr, loopKey) {
  expr = trim(expr)
  var present = {}
  var keys = []
  if (/\bmod\b/.test(expr) || /\bSUPER\b/.test(expr)) present.super = true

  var strings = []
  expr.replace(/"([^"]*)"/g, function (_, inner) {
    strings.push(inner)
    return ""
  })
  var glued = strings.join("")
  if (/\.\.\s*key\b/.test(expr) && loopKey != null) {
    glued += (glued ? " + " : "") + String(loopKey)
  }

  var tokens = glued.split("+")
  for (var i = 0; i < tokens.length; i++) {
    var tok = trim(tokens[i])
    if (!tok) continue
    var alias = MOD_ALIASES[tok.toLowerCase()]
    if (alias) {
      present[alias] = true
      continue
    }
    keys.push(tok)
  }
  if (keys.length === 0) keys.push("?")
  return { mods: formatMods(present), keys: keys }
}

function extractDesc(opts, loopI) {
  if (!opts) return null
  var m = opts.match(/desc\s*=\s*"([^"]*)"(\s*\.\.\s*(\w+))?/)
  if (!m) return null
  var text = m[1]
  if (m[3] === "i" && loopI != null) text += String(loopI)
  return text
}

function splitTopArgs(s) {
  var args = []
  var buf = ""
  var depth = 0
  var quote = null
  for (var i = 0; i < s.length; i++) {
    var c = s[i]
    if (quote) {
      buf += c
      if (c === quote && s[i - 1] !== "\\") quote = null
      continue
    }
    if (c === '"' || c === "'") {
      quote = c
      buf += c
      continue
    }
    if (c === "(" || c === "{") depth++
    else if (c === ")" || c === "}") depth--
    if (c === "," && depth === 0) {
      args.push(trim(buf))
      buf = ""
      continue
    }
    buf += c
  }
  if (trim(buf)) args.push(trim(buf))
  return args
}

function matchingParen(text, openIdx) {
  var depth = 0
  var quote = null
  for (var i = openIdx; i < text.length; i++) {
    var c = text[i]
    if (quote) {
      if (c === quote && text[i - 1] !== "\\") quote = null
      continue
    }
    if (c === '"' || c === "'") {
      quote = c
      continue
    }
    if (c === "(") depth++
    else if (c === ")") {
      depth--
      if (depth === 0) return i
    }
  }
  return -1
}

function isDecoration(content) {
  return /^[#=\-\s]*$/.test(content)
}

function isDisabledBind(content) {
  return /\b(hl\.bind|media_bind)\s*\(/.test(content)
}

function isSectionComment(content) {
  if (!content || isDecoration(content) || isDisabledBind(content)) return false
  if (/^local\s/.test(content) || /^function\b/.test(content)) return false
  return true
}

function parseForHeader(line) {
  var m = trim(line).match(/^for\s+\w+\s*=\s*(\d+)\s*,\s*(\d+)\s+do\b/)
  if (!m) return null
  return { start: Number(m[1]), end: Number(m[2]) }
}

function parseCall(stmt, loop) {
  var bind = stmt.match(/^\s*(hl\.bind|media_bind)\s*\(/)
  if (!bind) return null
  var open = stmt.indexOf("(")
  var close = matchingParen(stmt, open)
  if (close < 0) return null
  var inner = stmt.slice(open + 1, close)
  var args = splitTopArgs(inner)
  if (args.length < 1) return null
  var opts = args.length >= 3 ? args[2] : args[args.length - 1]
  if (opts && opts.charAt(0) !== "{") opts = ""
  var action = args.length >= 2 ? args[1] : ""

  var iterations = [null]
  if (loop && /\.\.\s*(key|i)\b/.test(args[0] + (opts || ""))) {
    iterations = []
    for (var n = loop.start; n <= loop.end; n++) iterations.push(n)
  }

  var out = []
  for (var i = 0; i < iterations.length; i++) {
    var iter = iterations[i]
    var loopKey = iter == null ? null : String(iter % 10)
    var chordInfo = parseChord(args[0], loopKey)
    var desc = extractDesc(opts || "", iter)
    out.push({
      mods: chordInfo.mods,
      keys: chordInfo.keys.slice(),
      desc: desc,
      action: action
    })
  }
  return out
}

function directionPrefix(desc) {
  var m = String(desc || "").match(/^(.*)\s+(left|right|up|down)$/i)
  return m ? { prefix: m[1], dir: m[2].toLowerCase() } : null
}

function stemDesc(text) {
  return String(text || "").replace(/\s+\d+$/, "")
}

function buildRows(entries) {
  var groups = []
  for (var i = 0; i < entries.length; i++) {
    var e = entries[i]
    var text = stemDesc(e.desc ? prettyDesc(e.desc) : deriveDesc(e.action))
    var prev = groups[groups.length - 1]
    var joins = prev && prev.mods === e.mods && prev.text === text
    if (joins) {
      for (var k = 0; k < e.keys.length; k++) prev.keys.push(e.keys[k])
    } else {
      groups.push({ mods: e.mods, keys: e.keys.slice(), text: text })
    }
  }

  var dirs = []
  for (var d = 0; d < groups.length; d++) {
    var g = groups[d]
    var info = directionPrefix(g.text)
    var last = dirs[dirs.length - 1]
    if (info && last && last.mods === g.mods && last.prefix === info.prefix) {
      for (var dk = 0; dk < g.keys.length; dk++) last.keys.push(g.keys[dk])
    } else if (info) {
      dirs.push({
        mods: g.mods,
        keys: g.keys.slice(),
        text: info.prefix,
        prefix: info.prefix,
        directional: true
      })
    } else {
      dirs.push({ mods: g.mods, keys: g.keys.slice(), text: g.text })
    }
  }

  var merged = []
  for (var m = 0; m < dirs.length; m++) {
    var cur = dirs[m]
    var found = null
    for (var p = 0; p < merged.length; p++) {
      if (merged[p].mods === cur.mods && merged[p].text === cur.text) {
        found = merged[p]
        break
      }
    }
    if (found) {
      for (var ck = 0; ck < cur.keys.length; ck++) found.keys.push(cur.keys[ck])
    } else {
      merged.push({ mods: cur.mods, keys: cur.keys.slice(), text: cur.text })
    }
  }

  var rows = []
  for (var r = 0; r < merged.length; r++) {
    var row = merged[r]
    var uniq = []
    var seen = {}
    for (var u = 0; u < row.keys.length; u++) {
      if (seen[row.keys[u]]) continue
      seen[row.keys[u]] = true
      uniq.push(row.keys[u])
    }
    rows.push({ desc: row.text, keys: chord(row.mods, uniq) })
  }
  return rows
}

function parse(text) {
  var raw = String(text || "").replace(/\r/g, "")
  var lines = raw.split("\n")
  var sections = []
  var current = { title: "General", entries: [] }
  var loop = null
  var loopDepth = 0
  var fnDepth = 0
  var pending = ""
  var i = 0

  function flush() {
    if (current.entries.length > 0) sections.push(current)
  }

  while (i < lines.length) {
    var line = lines[i]
    var trimmed = trim(line)

    if (trimmed === "") {
      i++
      continue
    }

    if (trimmed.indexOf("--") === 0) {
      var content = trim(trimmed.replace(/^--\s?/, ""))
      if (isSectionComment(content)) {
        flush()
        current = { title: capitalize(content), entries: [] }
      }
      i++
      continue
    }

    var forHead = parseForHeader(trimmed)
    if (forHead) {
      loop = forHead
      loopDepth = 1
      i++
      continue
    }

    if (/^\s*(local\s+)?function\b/.test(trimmed)) {
      fnDepth++
      i++
      continue
    }

    if (/^\s*end\b/.test(line)) {
      if (loop && loopDepth > 0) {
        loopDepth--
        if (loopDepth <= 0) loop = null
      } else if (fnDepth > 0) {
        fnDepth--
      }
      i++
      continue
    }

    if (fnDepth > 0) {
      i++
      continue
    }

    if (/^\s*(hl\.bind|media_bind)\s*\(/.test(line) || pending) {
      pending = pending ? pending + "\n" + line : line
      var open = pending.indexOf("(")
      var close = open >= 0 ? matchingParen(pending, open) : -1
      if (close < 0) {
        i++
        continue
      }
      var stmt = pending
      pending = ""
      var parsed = parseCall(stmt, loop)
      if (parsed) {
        for (var p = 0; p < parsed.length; p++) current.entries.push(parsed[p])
      }
      i++
      continue
    }

    i++
  }
  flush()

  var out = []
  for (var s = 0; s < sections.length; s++) {
    var rows = buildRows(sections[s].entries)
    if (rows.length > 0) out.push({ title: sections[s].title, rows: rows })
  }
  return out
}

function flatten(sections) {
  var elems = []
  for (var si = 0; si < sections.length; si++) {
    var sec = sections[si]
    elems.push({
      kind: "header",
      text: String(sec.title || "").toUpperCase(),
      h: 20,
      gap: si > 0 ? 10 : 0
    })
    var rows = sec.rows || []
    for (var r = 0; r < rows.length; r++) {
      elems.push({
        kind: "row",
        desc: rows[r].desc,
        keys: rows[r].keys,
        h: 19,
        gap: 0
      })
    }
  }
  return elems
}

function colHeight(elems, from, to) {
  var h = 0
  for (var i = from; i <= to; i++) {
    h += elems[i].h + (i > from ? elems[i].gap : 0)
  }
  return h
}

function layoutColumns(sections) {
  var elems = flatten(sections)
  var total = 0
  for (var i = 0; i < elems.length; i++) total += elems[i].gap + elems[i].h
  var split = elems.length
  var best = null
  var acc = 0
  for (var e = 0; e < elems.length; e++) {
    if (elems[e].kind === "header" && e > 0) {
      var d = Math.abs(acc - total / 2)
      if (best == null || d < best) {
        best = d
        split = e
      }
    }
    acc += elems[e].gap + elems[e].h
  }
  return {
    left: elems.slice(0, split),
    right: elems.slice(split),
    source: "hypr",
    height: Math.max(colHeight(elems, 0, split - 1), colHeight(elems, split, elems.length - 1))
  }
}

if (typeof module !== "undefined" && module.exports) {
  module.exports = {
    prettyDesc: prettyDesc,
    deriveDesc: deriveDesc,
    parseChord: parseChord,
    displayKeys: displayKeys,
    parse: parse,
    layoutColumns: layoutColumns
  }
}
